lib/block.ml

```diff
@@ -295,105 +296,195 @@ let version_enabled version =
 
 let get_label f (labels : Label.t list) = Util.List.find_map f labels
 
-let check_not_set msg = function
-  | Some _ -> Util.Result.errorf msg
-  | None -> Ok ()
+let label_not_allowed ~label ~kind =
+  Util.Result.errorf "`%s` label is not allowed for %s blocks." label kind
+
+let label_required ~label ~kind =
+  Util.Result.errorf "`%s` label is required for %s blocks." label kind
+
+let fmt_header = Fmt.using (Format.asprintf "`%a`" Header.pp) Fmt.string
+
+let header_not_allowed ~header ~kind =
```

Comment:
I wonder if we should really forbid any headers. When the block type is set explicitly, the header shouldn't matter. When the block type isn't set explicitly, we use it to determine the block type so this will never happen.

What do you think?

Reply:
I think it's better for consistency to forbid headers incompatible for blocks of type T, when said headers are incompatible with T. This way, when we remove the block type constraint the behavior of the block would be the same.

Reply:
The block header doesn't have an actual semantic value, it's just used for styling and in our case, to infer the block type when that's the only way. They're not incompatible, they just mess with our inference logic. If we set their type explicitly, mdx shouldn't try to reason about them as they are irrelevant to mdx. I feel like it's a bit intrusive.

Imagine I want to apply a custom style to some OCaml blocks and to do so I set the block header to `custom-ocaml` and explicitly set the mdx block type to ocaml. With the current behaviour, mdx will fail to parse the block, telling me I can't use this header for this block type. Even though this scenario is a bit of a fiction and I don't of any such cases, I think it's possible as some md interpreters just use the block header to set a class in the generated html div.
In such a case it would be incredibly frustrating to have it refused by mdx for no substantial reason since the block header prevents in no way mdx from interpreting the block.

Reply:
I see, good point.

Reply:
Agreed, it makes more sense, I will simplify this part of the code.